File: app/src/main/java/ru/e6atb/chat/MarkdownRenderer.java
package ru.e6atb.chat;

import android.graphics.Typeface;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.style.ClickableSpan;
import android.text.style.StrikethroughSpan;
import android.text.style.StyleSpan;
import android.text.style.TypefaceSpan;
import android.view.View;

import java.util.Locale;

final class MarkdownRenderer {
	interface Callbacks {
		void copyCode(String code);
		void openUrl(String url);
		void openMention(String login);
		int linkColor();
	}

	private final Callbacks callbacks;

	MarkdownRenderer(Callbacks callbacks) {
		this.callbacks = callbacks;
	}

	CharSequence render(String value) {
		value = DisplayText.safe(value);
		if (value.length() == 0) return "";
		SpannableStringBuilder out = new SpannableStringBuilder();
		int i = 0;
		while (i < value.length()) {
			if (value.charAt(i) == '`') {
				int end = value.indexOf('`', i + 1);
				if (end > i + 1) {
					appendCodeSpan(out, value.substring(i + 1, end));
					i = end + 1;
					continue;
				}
			}
			int urlEnd = linkEnd(value, i);
			if (urlEnd > i) {
				String label = value.substring(i, urlEnd);
				String url = label.startsWith("www.") ? "https://" + label : label;
				appendUrlSpan(out, label, url);
				i = urlEnd;
				continue;
			}
			int mentionEnd = mentionEnd(value, i);
			if (mentionEnd > i) {
				String mention = value.substring(i + 1, mentionEnd).toLowerCase(Locale.US);
				appendMentionSpan(out, value.substring(i, mentionEnd), mention);
				i = mentionEnd;
				continue;
			}
			if (startsWith(value, i, "**")) {
				int end = value.indexOf("**", i + 2);
				if (end > i + 2) {
					appendStyleSpan(out, value.substring(i + 2, end), new StyleSpan(Typeface.BOLD));
					i = end + 2;
					continue;
				}
			}
			if (startsWith(value, i, "__")) {
				int end = value.indexOf("__", i + 2);
				if (end > i + 2) {
					appendStyleSpan(out, value.substring(i + 2, end), new StyleSpan(Typeface.BOLD));
					i = end + 2;
					continue;
				}
			}
			if (startsWith(value, i, "~~")) {
				int end = value.indexOf("~~", i + 2);
				if (end > i + 2) {
					appendStyleSpan(out, value.substring(i + 2, end), new StrikethroughSpan());
					i = end + 2;
					continue;
				}
			}
			if (value.charAt(i) == '*') {
				int end = value.indexOf('*', i + 1);
				if (end > i + 1) {
					appendStyleSpan(out, value.substring(i + 1, end), new StyleSpan(Typeface.ITALIC));
					i = end + 1;
					continue;
				}
			}
			if (value.charAt(i) == '_') {
				int end = value.indexOf('_', i + 1);
				if (end > i + 1) {
					appendStyleSpan(out, value.substring(i + 1, end), new StyleSpan(Typeface.ITALIC));
					i = end + 1;
					continue;
				}
			}
			int cp = value.codePointAt(i);
			out.append(value, i, i + Character.charCount(cp));
			i += Character.charCount(cp);
		}
		return out;
	}

	private boolean startsWith(String value, int offset, String prefix) {
		return offset + prefix.length() <= value.length() && value.startsWith(prefix, offset);
	}

	private int linkEnd(String value, int offset) {
		if (!startsWith(value, offset, "http://") &&
			!startsWith(value, offset, "https://") &&
			!startsWith(value, offset, "www.")) {
			return -1;
		}
		int end = offset;
		while (end < value.length() && !Character.isWhitespace(value.charAt(end))) {
			end++;
		}
		while (end > offset && ".,;:!?)]}".indexOf(value.charAt(end - 1)) >= 0) {
			end--;
		}
		return end > offset ? end : -1;
	}

	private int mentionEnd(String value, int offset) {
		if (value.charAt(offset) != '@') return -1;
		if (offset > 0) {
			char prev = value.charAt(offset - 1);
			if (Character.isLetterOrDigit(prev) || prev == '_' || prev == '.') return -1;
		}
		int end = offset + 1;
		while (end < value.length()) {
			char c = value.charAt(end);
			if (!Character.isLetterOrDigit(c) && c != '_' && c != '-' && c != '.') break;
			end++;
		}
		return end > offset + 1 ? end : -1;
	}

	private void appendStyleSpan(SpannableStringBuilder out, String value, Object span) {
		int start = out.length();
		out.append(value);
		out.setSpan(span, start, out.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
	}

	private void appendCodeSpan(SpannableStringBuilder out, final String value) {
		int start = out.length();
		out.append(value);
		out.setSpan(new TypefaceSpan("monospace"), start, out.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
		out.setSpan(new ClickableSpan() {
			@Override
			public void onClick(View widget) {
				if (callbacks != null) callbacks.copyCode(value);
			}

			@Override
			public void updateDrawState(TextPaint ds) {
				super.updateDrawState(ds);
				ds.setTypeface(Typeface.MONOSPACE);
				ds.setColor(linkColor());
				ds.setUnderlineText(false);
			}
		}, start, out.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
	}

	private void appendUrlSpan(SpannableStringBuilder out, String label, final String url) {
		int start = out.length();
		out.append(label);
		out.setSpan(new ClickableSpan() {
			@Override
			public void onClick(View widget) {
				if (callbacks != null) callbacks.openUrl(url);
			}

			@Override
			public void updateDrawState(TextPaint ds) {
				super.updateDrawState(ds);
				ds.setColor(linkColor());
				ds.setUnderlineText(true);
			}
		}, start, out.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
	}

	private void appendMentionSpan(SpannableStringBuilder out, String label, final String login) {
		int start = out.length();
		out.append(label);
		out.setSpan(new ClickableSpan() {
			@Override
			public void onClick(View widget) {
				if (callbacks != null) callbacks.openMention(login);
			}

			@Override
			public void updateDrawState(TextPaint ds) {
				super.updateDrawState(ds);
				ds.setColor(linkColor());
				ds.setUnderlineText(false);
			}
		}, start, out.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
	}

	private int linkColor() {
		return callbacks == null ? 0xff7fb4ff : callbacks.linkColor();
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/MarkdownRenderer.kt
package ru.e6atb.chat

import android.graphics.Typeface
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextPaint
import android.text.style.ClickableSpan
import android.text.style.StrikethroughSpan
import android.text.style.StyleSpan
import android.text.style.TypefaceSpan
import android.view.View
import java.util.Locale

internal class MarkdownRenderer(private val callbacks: Callbacks?) {
    interface Callbacks { fun copyCode(code: String?); fun openUrl(url: String?); fun openMention(login: String?); fun canRunBotCommand(): Boolean; fun runBotCommand(command: String?); fun linkColor(): Int }
    fun render(raw: String?): CharSequence {
        val value = DisplayText.safe(raw); if (value.isEmpty()) return ""
        val out = SpannableStringBuilder(); var index = 0
        while (index < value.length) {
            val codeEnd = if (value[index] == '`') value.indexOf('`', index + 1) else -1
            if (codeEnd > index + 1) { appendCodeSpan(out, value.substring(index + 1, codeEnd)); index = codeEnd + 1; continue }
            val urlEnd = linkEnd(value, index)
            if (urlEnd > index) { val label = value.substring(index, urlEnd); appendUrlSpan(out, label, if (label.startsWith("www.")) "https://$label" else label); index = urlEnd; continue }
            val mentionEnd = mentionEnd(value, index)
            if (mentionEnd > index) { appendMentionSpan(out, value.substring(index, mentionEnd), value.substring(index + 1, mentionEnd).lowercase(Locale.US)); index = mentionEnd; continue }
            val commandEnd = commandEnd(value, index)
            if (commandEnd > index && callbacks?.canRunBotCommand() == true) { appendCommandSpan(out, value.substring(index, commandEnd)); index = commandEnd; continue }
            val match = listOf("**" to StyleSpan(Typeface.BOLD), "__" to StyleSpan(Typeface.BOLD), "~~" to StrikethroughSpan()).firstOrNull { (prefix, _) -> startsWith(value, index, prefix) }
            if (match != null) { val end = value.indexOf(match.first, index + match.first.length); if (end > index + match.first.length) { appendStyleSpan(out, value.substring(index + match.first.length, end), match.second); index = end + match.first.length; continue } }
            if (value[index] == '*' || value[index] == '_') { val end = value.indexOf(value[index], index + 1); if (end > index + 1) { appendStyleSpan(out, value.substring(index + 1, end), StyleSpan(Typeface.ITALIC)); index = end + 1; continue } }
            val codePoint = value.codePointAt(index); out.append(value, index, index + Character.charCount(codePoint)); index += Character.charCount(codePoint)
        }
        return out
    }
    private fun startsWith(value: String, offset: Int, prefix: String) = offset + prefix.length <= value.length && value.startsWith(prefix, offset)
    private fun linkEnd(value: String, offset: Int): Int { if (!startsWith(value, offset, "http://") && !startsWith(value, offset, "https://") && !startsWith(value, offset, "www.")) return -1; var end = offset; while (end < value.length && !Character.isWhitespace(value[end])) end++; while (end > offset && ".,;:!?)]}".indexOf(value[end - 1]) >= 0) end--; return if (end > offset) end else -1 }
    private fun mentionEnd(value: String, offset: Int): Int { if (value[offset] != '@') return -1; if (offset > 0 && (Character.isLetterOrDigit(value[offset - 1]) || value[offset - 1] == '_' || value[offset - 1] == '.')) return -1; var end = offset + 1; while (end < value.length && (Character.isLetterOrDigit(value[end]) || value[end] == '_' || value[end] == '-' || value[end] == '.')) end++; return if (end > offset + 1) end else -1 }
    private fun commandEnd(value: String, offset: Int): Int { if (value[offset] != '/') return -1; if (offset > 0 && (Character.isLetterOrDigit(value[offset - 1]) || value[offset - 1] == '_' || value[offset - 1] == '-')) return -1; var end = offset + 1; while (end < value.length && (value[end].isLetterOrDigit() || value[end] == '_')) end++; return if (end > offset + 1 && end - offset <= 33) end else -1 }
    private fun appendStyleSpan(out: SpannableStringBuilder, value: String, span: Any) { val start = out.length; out.append(value); out.setSpan(span, start, out.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE) }
    private fun appendCodeSpan(out: SpannableStringBuilder, value: String) { val start = out.length; out.append(value); out.setSpan(TypefaceSpan("monospace"), start, out.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE); out.setSpan(span(value, false) { callbacks?.copyCode(value) }, start, out.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE) }
    private fun appendUrlSpan(out: SpannableStringBuilder, label: String, url: String) { val start = out.length; out.append(label); out.setSpan(span(url, true) { callbacks?.openUrl(url) }, start, out.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE) }
    private fun appendMentionSpan(out: SpannableStringBuilder, label: String, login: String) { val start = out.length; out.append(label); out.setSpan(span(login, false) { callbacks?.openMention(login) }, start, out.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE) }
    private fun appendCommandSpan(out: SpannableStringBuilder, command: String) { val start = out.length; out.append(command); out.setSpan(span(command, false) { callbacks?.runBotCommand(command) }, start, out.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE) }
    private fun span(value: String, underline: Boolean, click: () -> Unit) = object : ClickableSpan() { override fun onClick(widget: View) = click(); override fun updateDrawState(paint: TextPaint) { super.updateDrawState(paint); paint.color = callbacks?.linkColor() ?: 0xff7fb4ff.toInt(); paint.isUnderlineText = underline; if (!underline) paint.typeface = Typeface.MONOSPACE.takeIf { value.isNotEmpty() } ?: paint.typeface } }
}
